### ISSUE TITLE
Unbound local variable `index` referenced before assignment

### ISSUE THREAD

**Issue by gerrymanoim** [dask/dask#6954]
**What happened**:

`UnboundLocalError: local variable 'index' referenced before assignment` originating from an unhandled case in https://github.com/dask/dask/blob/master/dask/dataframe/core.py#L2805-L2812. 

**What you expected to happen**:

A `RuntimeError`? Some other sort of exception. 

**Minimal Complete Verifiable Example**:

```python
import dask.dataframe as dd
import numpy as np

s = dd.from_array(np.arange(10))
np.repeat(s, 10)
```

**Anything else we need to know?**:

I don't expect `np.repeat(s, 10)` to actually work (I triggered this case debugging some broken code), but was surprised that it resulted in an `UnboundLocalError`. 

**Environment**:

- Dask version: 2.30.0
- Python version: 3.8.5
- Operating System: Ubuntu 18.04
- Install method (conda, pip, source): conda

**Comment by jsignell** [comment]
Thanks for raising this! I can reproduce and see the issue. I think this is expected to work because of [NEP 18](https://numpy.org/neps/nep-0018-array-function-protocol.html). I am looking into it now.

**Comment by jsignell** [comment]
I somehow missed that this is a series being passed in. I agree that that is not expected to work because the method still needs to be implemented on the dask side, which `repeat` isn't for dataframe/series. 

You might want to use dask.array:

```python
import dask.array as da
import numpy as np

a = da.from_array(np.arange(10))
np.repeat(a, 10)
```

**Comment by gerrymanoim** [comment]
Right - I fully agree that this isn't the "correct" way to do anything (I hit this accidentally). I was just surprised at the failure mode.

**Comment by jsignell** [comment]
Yep :+1: on the surprise, I am trying to figure out what the behavior should be.

**Comment by jsignell** [comment]
@jrbourbeau I am wondering if you would be able to comment on what the behavior should be. I tried adding:

```python
if context is None:
    index = None
```

But the results in a pandas series being the output of `np.repeat(s)` which is definitely not expected. Poking around, I suspect that we need to do something before `__array_wrap__` is called, since in that method, `array` is already computed.
